nova/compute/manager.py

```diff
@@ -5851,6 +5859,176 @@ def _complete_volume_attachments(self, context, bdms):
                     self.volume_api.attachment_complete(
                         context, bdm.attachment_id)
 
+    def _update_ephemeral_bdms(self, context, instance, block_device_info,
+                               ephemeral_size, swap_size):
+        """Update ephemeral bdm objects in DB (for instance resizing).
+
+        Updates volume size of block device mapping object according to
+        specified new size. Works only for a single ephemeral and swap disks.
+        Stores the result into the DB.
+
+        Creates a new bdm if necessary.
+
+        :param context: Request context object.
+        :param instance: An instance which bdms are to be updated.
+        :param block_device_info: Instance block device mapping in driver
+                                  format.
+        :param ephemeral_size: New size in GB of the single ephemeral disk.
+                               Creates a new bdm if no bdm exists.
+                               If None - does nothing with ephemerals.
+                               If 0 - removes the ephemeral disk
+        :param swap_size: A size in MB of swap disk. Creates a new bdm if no
+                          swap bdm exists and non zero size is specified.
+                          Destroys swap bdm if zero size is specified.
+                          If None - does nothing with swap.
+        """
+        ephemerals = driver.block_device_info_get_ephemerals(block_device_info)
+        swap = driver.block_device_info_get_swap(block_device_info)
+        reassing_device_names = False
+
+        # Resize ephemeral disk
+        if ephemeral_size is not None:
+            if not ephemerals and ephemeral_size > 0:
+                LOG.debug(
+                    f'Creating ephemeral bdm: size={ephemeral_size}.',
+                    instance=instance
+                )
+                bdm = objects.BlockDeviceMapping(
+                    context=context, instance_uuid=instance.uuid,
+                    source_type='blank', destination_type='local',
+                    device_type='disk', delete_on_termination=True,
+                    boot_index=-1, volume_size=ephemeral_size)
+                bdm.create()
+                reassing_device_names = True
+            elif len(ephemerals) == 1 and ephemeral_size == 0:
+                LOG.debug('Deleting ephemeral bdm.', instance=instance)
+                ephemerals[0].destroy()
+                reassing_device_names = True
+            elif len(ephemerals) == 1:
+                LOG.debug(
+                    f'Resizing ephemeral bdm: new_size={ephemeral_size}, '
+                    f'old_size={ephemerals[0].volume_size}.',
+                    instance=instance,
+                )
+                ephemerals[0].volume_size = ephemeral_size
```

Comment:
Might it make sense to assert that the volume size not smaller than before here?